socketio/src/client/reconnect.rs

```diff
@@ -0,0 +1,239 @@
+#![allow(unused)]
+use std::{
+    ops::Deref,
+    sync::{Arc, Mutex, RwLock},
+    time::Duration,
+};
+
+use super::{Client, ClientBuilder};
+use crate::{error::Result, packet::Packet, Error};
+pub use crate::{event::Event, payload::Payload};
+use backoff::ExponentialBackoff;
+use backoff::{backoff::Backoff, ExponentialBackoffBuilder};
+
+#[derive(Clone)]
+pub struct ReconnectClient {
+    builder: ClientBuilder,
+    client: Arc<RwLock<Client>>,
+    backoff: ExponentialBackoff,
+}
+
+impl ReconnectClient {
+    pub(crate) fn new(builder: ClientBuilder) -> Result<Self> {
+        let builder_clone = builder.clone();
+        let client = builder_clone.connect_manual()?;
+        let backoff = ExponentialBackoffBuilder::new()
+            .with_initial_interval(Duration::from_millis(builder.reconnect_delay_min))
+            .with_max_interval(Duration::from_millis(builder.reconnect_delay_max))
+            .build();
+
+        let s = Self {
+            builder,
+            client: Arc::new(RwLock::new(client)),
+            backoff: ExponentialBackoff::default(),
+        };
+        s.poll_callback();
+
+        Ok(s)
+    }
+
+    pub fn emit<E, D>(&self, event: E, data: D) -> Result<()>
+    where
+        E: Into<Event>,
+        D: Into<Payload>,
+    {
+        let client = self.client.read()?;
+        // TODO(#230): like js client, buffer emit, resend after reconnect
+        client.emit(event, data)
+    }
+
+    pub fn emit_with_ack<F, E, D>(
+        &self,
+        event: E,
+        data: D,
+        timeout: Duration,
+        callback: F,
+    ) -> Result<()>
+    where
+        F: for<'a> FnMut(Payload, Client) + 'static + Sync + Send,
+        E: Into<Event>,
+        D: Into<Payload>,
+    {
+        let client = self.client.read()?;
+        // TODO(#230): like js client, buffer emit, resend after reconnect
+        client.emit_with_ack(event, data, timeout, callback)
+    }
+
+    pub fn disconnect(&self) -> Result<()> {
+        let client = self.client.read()?;
+        client.disconnect()
+    }
+
+    fn reconnect(&mut self) -> Result<()> {
+        let mut reconnect_attempts = 0;
+        if self.builder.reconnect {
```

Comment:
I checked js client, they can set `reconnect = false` to disable reconnect. we can leave `connect` API to normal Client, but add a new API like `reconnectable_connect`(not a good name) to get rid of `reconnect: bool` variable.

Reply:
(another option)
solution three: rename ReconnctingClient to Client (and take it's place), and rename existing client to InternalClient

Thoughts @1c3t3a

Reply:
I agree with you @nshaaban-cPacket, this makes sense. Changing the Clients default reconnection behavior is a breaking change anyways, so it would make sense to release a new breaking version and be transparent about what has changed.